<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>CellChat example</title>

<script src="site_libs/header-attrs-2.1/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
#rmd-source-code {
  display: none;
}
</style>

<link rel="stylesheet" href="custom.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Interaction tools</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/lazappi/interaction-tools">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
<li role="separator" class="divider"></li>
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">CellChat example</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-07-30
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>interaction-tools/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191213code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191213)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191213code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191213)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpgithubcomlazappiiteractiontoolstreef01744803e8182dfa5356d3f38543fed78292e13targetblankf017448a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="http://github.com/lazappi/iteraction-tools/tree/f01744803e8182dfa5356d3f38543fed78292e13" target="_blank">f017448</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpgithubcomlazappiiteractiontoolstreef01744803e8182dfa5356d3f38543fed78292e13targetblankf017448a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="http://github.com/lazappi/iteraction-tools/tree/f01744803e8182dfa5356d3f38543fed78292e13" target="_blank">f017448</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .drake/
    Ignored:    data/COMUNET/
    Ignored:    data/CellChat/
    Ignored:    data/NicheNet/
    Ignored:    data/cellphonedb/
    Ignored:    output/14-CellChat.Rmd/
    Ignored:    output/index.Rmd/
    Ignored:    renv/library/
    Ignored:    renv/python/
    Ignored:    renv/staging/

Unstaged changes:
    Modified:   analysis/14-CellChat.Rmd
    Modified:   code/setup.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/14-CellChat.Rmd</code>) and HTML (<code>docs/14-CellChat.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="http://github.com/lazappi/iteraction-tools/blob/f01744803e8182dfa5356d3f38543fed78292e13/analysis/14-CellChat.Rmd" target="_blank">f017448</a>
</td>
<td>
Luke Zappia
</td>
<td>
2020-07-29
</td>
<td>
Setup CellChat
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="http://github.com/lazappi/iteraction-tools/blob/f01744803e8182dfa5356d3f38543fed78292e13/docs/14-CellChat.html" target="_blank">f017448</a>
</td>
<td>
Luke Zappia
</td>
<td>
2020-07-29
</td>
<td>
Setup CellChat
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code># Setup document
source(here::here(&quot;code&quot;, &quot;setup.R&quot;))

# Function dependencies
invisible(drake::readd(download_link))</code></pre>
<div id="introduction" class="section level1 unnumbered" number="">
<h1 class="unnumbered" number="">Introduction</h1>
<p>In this document we are going to run through the example analysis for the <strong>CellChat</strong> package and have a look at the output it produces. More information about <strong>CellChat</strong> can be found at <a href="https://github.com/sqjin/CellChat" class="uri">https://github.com/sqjin/CellChat</a>.</p>
<pre class="r"><code>library(&quot;CellChat&quot;)
library(&quot;ggalluvial&quot;)</code></pre>
<p class="timeit">
Chunk time: 2.92 secs
</p>
</div>
<div id="input" class="section level1" number="1">
<h1 number="1"><span class="header-section-number">1</span> Input</h1>
<p><strong>CellChat</strong> requires two user inputs: an expression matrix and either cell-type labels or a low-dimensional embedding. Here we use the label-based mode. The other inputs to <strong>CellChat</strong> are the ligand-receptor database and protein-protein interaction network included in the package.</p>
<pre class="r"><code>load(fs::path(PATHS$CellChat_in, &quot;data_wound_CellChat.rda&quot;))</code></pre>
<p class="timeit">
Chunk time: 2 secs
</p>
<div id="expression-matrix" class="section level2" number="1.1">
<h2 number="1.1"><span class="header-section-number">1.1</span> Expression matrix</h2>
<p>The expression matrix should be normalised with genes as rows and cells as columns. Here is a snippet of the example dataset:</p>
<pre class="r"><code>expr &lt;- data_wound$data

pander(as.matrix(expr[1:5, 1:5]))</code></pre>
<table style="width:97%;">
<caption>Table continues below</caption>
<colgroup>
<col width="18%" />
<col width="26%" />
<col width="26%" />
<col width="26%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">AAACCTGAGATGTGGC</th>
<th align="center">AAACCTGAGGTGTTAA</th>
<th align="center">AAACCTGAGTATCTCG</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>Xkr4</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>Sox17</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>Mrpl15</strong></td>
<td align="center">0</td>
<td align="center">1.724</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>Lypla1</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>Tcea1</strong></td>
<td align="center">1.432</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<table style="width:71%;">
<colgroup>
<col width="18%" />
<col width="26%" />
<col width="26%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">AAACCTGCAAAGCAAT</th>
<th align="center">AAACCTGCAGCTTCGG</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>Xkr4</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>Sox17</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>Mrpl15</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>Lypla1</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>Tcea1</strong></td>
<td align="center">1.418</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p class="timeit">
Chunk time: 0.16 secs
</p>
<p>The full matrix has <strong>17090</strong> rows (genes) and <strong>21557</strong> columns (cells).</p>
</div>
<div id="labels" class="section level2" number="1.2">
<h2 number="1.2"><span class="header-section-number">1.2</span> Labels</h2>
<p>The second input is a <code>data.frame</code> with cell type labels for each cell.</p>
<pre class="r"><code>labels &lt;- data.frame(
    group     = data_wound$labels,
    row.names = names(data_wound$labels)
)
group_size &lt;- as.numeric(table(labels$group))

skim(labels)</code></pre>
<table>
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td align="left">Name</td>
<td align="left">labels</td>
</tr>
<tr class="even">
<td align="left">Number of rows</td>
<td align="left">21557</td>
</tr>
<tr class="odd">
<td align="left">Number of columns</td>
<td align="left">1</td>
</tr>
<tr class="even">
<td align="left">_______________________</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Column type frequency:</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">factor</td>
<td align="left">1</td>
</tr>
<tr class="odd">
<td align="left">________________________</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">Group variables</td>
<td align="left">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: factor</strong></p>
<table>
<thead>
<tr class="header">
<th align="left">skim_variable</th>
<th align="right">n_missing</th>
<th align="right">complete_rate</th>
<th align="left">ordered</th>
<th align="right">n_unique</th>
<th align="left">top_counts</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">group</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">25</td>
<td align="left">FIB: 5073, FIB: 2817, FIB: 1957, FIB: 1323</td>
</tr>
</tbody>
</table>
<p class="timeit">
Chunk time: 0.07 secs
</p>
</div>
<div id="database" class="section level2" number="1.3">
<h2 number="1.3"><span class="header-section-number">1.3</span> Database</h2>
<p>The database is available as a list. The tutorial suggests using only the interactions annotated with “Secreted Signalling”.</p>
<pre class="r"><code>CellChatDB &lt;- CellChatDB.mouse
CellChatDB &lt;- subsetDB(CellChatDB, search = &quot;Secreted Signaling&quot;)</code></pre>
<p class="timeit">
Chunk time: 0.18 secs
</p>
<p>The database list contains <strong>4</strong> items: <strong>interaction</strong>, <strong>complex</strong>, <strong>cofactor</strong> and <strong>geneInfo</strong></p>
<div id="interaction" class="section level3" number="1.3.1">
<h3 number="1.3.1"><span class="header-section-number">1.3.1</span> Interaction</h3>
<p>The first item in the list is a <code>data.frame</code> with information about interactions.</p>
<pre class="r"><code>skim(CellChatDB$interaction)</code></pre>
<table>
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td align="left">Name</td>
<td align="left">CellChatDB$interaction</td>
</tr>
<tr class="even">
<td align="left">Number of rows</td>
<td align="left">1211</td>
</tr>
<tr class="odd">
<td align="left">Number of columns</td>
<td align="left">11</td>
</tr>
<tr class="even">
<td align="left">_______________________</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Column type frequency:</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">character</td>
<td align="left">11</td>
</tr>
<tr class="odd">
<td align="left">________________________</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">Group variables</td>
<td align="left">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: character</strong></p>
<table>
<thead>
<tr class="header">
<th align="left">skim_variable</th>
<th align="right">n_missing</th>
<th align="right">complete_rate</th>
<th align="right">min</th>
<th align="right">max</th>
<th align="right">empty</th>
<th align="right">n_unique</th>
<th align="right">whitespace</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">interaction_name</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">6</td>
<td align="right">21</td>
<td align="right">0</td>
<td align="right">1211</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">pathway_name</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">12</td>
<td align="right">0</td>
<td align="right">141</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">ligand</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">12</td>
<td align="right">0</td>
<td align="right">361</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">receptor</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">3</td>
<td align="right">18</td>
<td align="right">0</td>
<td align="right">363</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">agonist</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">15</td>
<td align="right">868</td>
<td align="right">5</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">antagonist</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">18</td>
<td align="right">755</td>
<td align="right">11</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">co_A_receptor</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">25</td>
<td align="right">829</td>
<td align="right">9</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">co_I_receptor</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">27</td>
<td align="right">724</td>
<td align="right">10</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">evidence</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">12</td>
<td align="right">45</td>
<td align="right">0</td>
<td align="right">376</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">annotation</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">18</td>
<td align="right">18</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">interaction_name_2</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">9</td>
<td align="right">26</td>
<td align="right">0</td>
<td align="right">1211</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p class="timeit">
Chunk time: 0.09 secs
</p>
</div>
<div id="complex" class="section level3" number="1.3.2">
<h3 number="1.3.2"><span class="header-section-number">1.3.2</span> Complex</h3>
<p>The second item contain information about complexes.</p>
<pre class="r"><code>skim(CellChatDB$complex)</code></pre>
<table>
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td align="left">Name</td>
<td align="left">CellChatDB$complex</td>
</tr>
<tr class="even">
<td align="left">Number of rows</td>
<td align="left">155</td>
</tr>
<tr class="odd">
<td align="left">Number of columns</td>
<td align="left">4</td>
</tr>
<tr class="even">
<td align="left">_______________________</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Column type frequency:</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">character</td>
<td align="left">4</td>
</tr>
<tr class="odd">
<td align="left">________________________</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">Group variables</td>
<td align="left">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: character</strong></p>
<table>
<thead>
<tr class="header">
<th align="left">skim_variable</th>
<th align="right">n_missing</th>
<th align="right">complete_rate</th>
<th align="right">min</th>
<th align="right">max</th>
<th align="right">empty</th>
<th align="right">n_unique</th>
<th align="right">whitespace</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">subunit_1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">3</td>
<td align="right">7</td>
<td align="right">0</td>
<td align="right">96</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">subunit_2</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">3</td>
<td align="right">7</td>
<td align="right">0</td>
<td align="right">70</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">subunit_3</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="right">142</td>
<td align="right">10</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">subunit_4</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">154</td>
<td align="right">2</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p class="timeit">
Chunk time: 0.06 secs
</p>
</div>
<div id="cofactor" class="section level3" number="1.3.3">
<h3 number="1.3.3"><span class="header-section-number">1.3.3</span> Cofactor</h3>
<p>The next item contains information about cofactors.</p>
<pre class="r"><code>skim(CellChatDB$cofactor)</code></pre>
<table>
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td align="left">Name</td>
<td align="left">CellChatDB$cofactor</td>
</tr>
<tr class="even">
<td align="left">Number of rows</td>
<td align="left">31</td>
</tr>
<tr class="odd">
<td align="left">Number of columns</td>
<td align="left">16</td>
</tr>
<tr class="even">
<td align="left">_______________________</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Column type frequency:</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">character</td>
<td align="left">16</td>
</tr>
<tr class="odd">
<td align="left">________________________</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">Group variables</td>
<td align="left">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: character</strong></p>
<table>
<thead>
<tr class="header">
<th align="left">skim_variable</th>
<th align="right">n_missing</th>
<th align="right">complete_rate</th>
<th align="right">min</th>
<th align="right">max</th>
<th align="right">empty</th>
<th align="right">n_unique</th>
<th align="right">whitespace</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">cofactor1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">6</td>
<td align="right">0</td>
<td align="right">25</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cofactor2</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="right">19</td>
<td align="right">12</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">cofactor3</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="right">22</td>
<td align="right">9</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cofactor4</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">8</td>
<td align="right">24</td>
<td align="right">7</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">cofactor5</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="right">25</td>
<td align="right">6</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cofactor6</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="right">26</td>
<td align="right">5</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">cofactor7</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="right">28</td>
<td align="right">4</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cofactor8</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="right">28</td>
<td align="right">4</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">cofactor9</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">29</td>
<td align="right">3</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cofactor10</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">7</td>
<td align="right">29</td>
<td align="right">3</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">cofactor11</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">30</td>
<td align="right">2</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cofactor12</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">30</td>
<td align="right">2</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">cofactor13</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="right">30</td>
<td align="right">2</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cofactor14</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="right">30</td>
<td align="right">2</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">cofactor15</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">7</td>
<td align="right">30</td>
<td align="right">2</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cofactor16</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">30</td>
<td align="right">2</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p class="timeit">
Chunk time: 0.12 secs
</p>
</div>
<div id="gene-info" class="section level3" number="1.3.4">
<h3 number="1.3.4"><span class="header-section-number">1.3.4</span> Gene info</h3>
<p>The final item in the database is gene annotation information.</p>
<pre class="r"><code>skim(CellChatDB$geneInfo)</code></pre>
<table>
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td align="left">Name</td>
<td align="left">CellChatDB$geneInfo</td>
</tr>
<tr class="even">
<td align="left">Number of rows</td>
<td align="left">45544</td>
</tr>
<tr class="odd">
<td align="left">Number of columns</td>
<td align="left">6</td>
</tr>
<tr class="even">
<td align="left">_______________________</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Column type frequency:</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">character</td>
<td align="left">6</td>
</tr>
<tr class="odd">
<td align="left">________________________</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">Group variables</td>
<td align="left">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: character</strong></p>
<table>
<thead>
<tr class="header">
<th align="left">skim_variable</th>
<th align="right">n_missing</th>
<th align="right">complete_rate</th>
<th align="right">min</th>
<th align="right">max</th>
<th align="right">empty</th>
<th align="right">n_unique</th>
<th align="right">whitespace</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Symbol</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">16</td>
<td align="right">0</td>
<td align="right">45544</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">Name</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">5</td>
<td align="right">139</td>
<td align="right">0</td>
<td align="right">45528</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">EntrezGene.ID</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">9</td>
<td align="right">0</td>
<td align="right">45536</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">Ensembl.Gene.ID</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">18</td>
<td align="right">5</td>
<td align="right">36640</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">HomoloGene.ID</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="right">4</td>
<td align="right">18941</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">HGNC.ID</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">85</td>
<td align="right">2</td>
<td align="right">17932</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p class="timeit">
Chunk time: 0.27 secs
</p>
</div>
</div>
<div id="network" class="section level2" number="1.4">
<h2 number="1.4"><span class="header-section-number">1.4</span> Network</h2>
<p>The network is a binary matrix showing interactions between proteins.</p>
<pre class="r"><code>pander(as.matrix(PPI.mouse[1:5, 1:5]))</code></pre>
<table style="width:69%;">
<colgroup>
<col width="16%" />
<col width="9%" />
<col width="11%" />
<col width="11%" />
<col width="9%" />
<col width="11%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">Cdh1</th>
<th align="center">Bcl6b</th>
<th align="center">Pparg</th>
<th align="center">Raf1</th>
<th align="center">Kat2b</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>Cdh1</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>Bcl6b</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>Pparg</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>Raf1</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>Kat2b</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p class="timeit">
Chunk time: 0.01 secs
</p>
<p>The full matrix has <strong>557</strong> rows and <strong>557</strong> columns.</p>
</div>
</div>
<div id="create-cellchat-object" class="section level1" number="2">
<h1 number="2"><span class="header-section-number">2</span> Create <code>CellChat</code> object</h1>
<p><strong>CellChat</strong> uses a custom object which we can create from the expression matrix and labels. We also store the database here.</p>
<pre class="r"><code>cellchat &lt;- createCellChat(expr)
cellchat &lt;- addMeta(cellchat, meta = labels, meta.name = &quot;labels&quot;)
cellchat &lt;- setIdent(cellchat, ident.use = &quot;labels&quot;)
cellchat@DB &lt;- CellChatDB

cellchat</code></pre>
<pre><code>An object of class CellChat 
 17090 genes.
 21557 cells.</code></pre>
<p class="timeit">
Chunk time: 0.01 secs
</p>
</div>
<div id="pre-processing" class="section level1" number="3">
<h1 number="3"><span class="header-section-number">3</span> Pre-processing</h1>
<div id="subset-signalling-genes" class="section level2" number="3.1">
<h2 number="3.1"><span class="header-section-number">3.1</span> Subset signalling genes</h2>
<p>First we subset the object to select only the signalling genes that are listed in the database.</p>
<pre class="r"><code>cellchat &lt;- subsetData(cellchat)

cellchat</code></pre>
<pre><code>An object of class CellChat 
 17090 genes.
 21557 cells.</code></pre>
<p class="timeit">
Chunk time: 0.22 secs
</p>
</div>
<div id="identify-over-expression" class="section level2" number="3.2">
<h2 number="3.2"><span class="header-section-number">3.2</span> Identify over-expression</h2>
<p>The next step is to identify over-expressed genes and interactions. Over-expressed genes are detected using a one-sided Wilcoxon test. Interactions are selected if either a ligand or receptor is over-expressed.</p>
<pre class="r"><code>cellchat &lt;- identifyOverExpressedGenes(cellchat)
cellchat &lt;- identifyOverExpressedInteractions(cellchat)</code></pre>
<p class="timeit">
Chunk time: 8.53 mins
</p>
<p>This step has identified <strong>539</strong> genes and <strong>666</strong> interactions.</p>
</div>
<div id="project-data" class="section level2" number="3.3">
<h2 number="3.3"><span class="header-section-number">3.3</span> Project data</h2>
<p>The gene expression data is then projected onto a protein-protein interaction network.</p>
<pre class="r"><code>cellchat &lt;- projectData(cellchat, PPI.mouse)</code></pre>
<p class="timeit">
Chunk time: 1.78 secs
</p>
</div>
</div>
<div id="communication-network-inference" class="section level1" number="4">
<h1 number="4"><span class="header-section-number">4</span> Communication network inference</h1>
<div id="compute-communication-probability" class="section level2" number="4.1">
<h2 number="4.1"><span class="header-section-number">4.1</span> Compute communication probability</h2>
<p>At this stage we can now compute the communication probability and infer the communication network. The results are stored in the <code>net</code> slot of the <code>CellChat</code> object which contains three-dimensional arrays of probabilities and p-values for interactions between each pair of cell types for each interaction pair. Here are snippets of these arrays for the first interaction pair.</p>
<pre class="r"><code>cellchat &lt;- computeCommunProb(cellchat)

pander(cellchat@net$prob[1:5, 1:5, 1])</code></pre>
<table style="width:89%;">
<colgroup>
<col width="16%" />
<col width="11%" />
<col width="11%" />
<col width="16%" />
<col width="16%" />
<col width="16%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">FIB-A</th>
<th align="center">FIB-B</th>
<th align="center">FIB-C</th>
<th align="center">FIB-D</th>
<th align="center">FIB-E</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>FIB-A</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">1.186e-07</td>
<td align="center">1.152e-06</td>
<td align="center">5.496e-07</td>
</tr>
<tr class="even">
<td align="center"><strong>FIB-B</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>FIB-C</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">8.06e-08</td>
<td align="center">7.834e-07</td>
<td align="center">3.737e-07</td>
</tr>
<tr class="even">
<td align="center"><strong>FIB-D</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">7.92e-08</td>
<td align="center">7.707e-07</td>
<td align="center">3.676e-07</td>
</tr>
<tr class="odd">
<td align="center"><strong>FIB-E</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">5.006e-08</td>
<td align="center">4.868e-07</td>
<td align="center">2.322e-07</td>
</tr>
</tbody>
</table>
<pre class="r"><code>pander(cellchat@net$pval[1:5, 1:5, 1])</code></pre>
<table style="width:72%;">
<colgroup>
<col width="16%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">FIB-A</th>
<th align="center">FIB-B</th>
<th align="center">FIB-C</th>
<th align="center">FIB-D</th>
<th align="center">FIB-E</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>FIB-A</strong></td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>FIB-B</strong></td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>
<tr class="odd">
<td align="center"><strong>FIB-C</strong></td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>FIB-D</strong></td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0.09</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>FIB-E</strong></td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0.22</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p class="timeit">
Chunk time: 5.03 mins
</p>
</div>
<div id="infer-pathway-signalling" class="section level2" number="4.2">
<h2 number="4.2"><span class="header-section-number">4.2</span> Infer pathway signalling</h2>
<p>We also compute the communication probabilities at the pathway level by summarising all ligand-receptor interactions associated with each pathway. The output is similar but now instead of interaction pairs we have pathways.</p>
<pre class="r"><code>cellchat &lt;- computeCommunProbPathway(cellchat)

pander(cellchat@netP$prob[1:5, 1:5, 1])</code></pre>
<table style="width:89%;">
<colgroup>
<col width="16%" />
<col width="11%" />
<col width="11%" />
<col width="16%" />
<col width="16%" />
<col width="16%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">FIB-A</th>
<th align="center">FIB-B</th>
<th align="center">FIB-C</th>
<th align="center">FIB-D</th>
<th align="center">FIB-E</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>FIB-A</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">1.186e-07</td>
<td align="center">1.152e-06</td>
<td align="center">5.496e-07</td>
</tr>
<tr class="even">
<td align="center"><strong>FIB-B</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>FIB-C</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">8.06e-08</td>
<td align="center">7.834e-07</td>
<td align="center">3.737e-07</td>
</tr>
<tr class="even">
<td align="center"><strong>FIB-D</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">8.283e-08</td>
<td align="center">1.577e-06</td>
<td align="center">7.521e-07</td>
</tr>
<tr class="odd">
<td align="center"><strong>FIB-E</strong></td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">5.867e-08</td>
<td align="center">1.057e-06</td>
<td align="center">5.043e-07</td>
</tr>
</tbody>
</table>
<p class="timeit">
Chunk time: 1.2 secs
</p>
</div>
<div id="aggregate-network" class="section level2" number="4.3">
<h2 number="4.3"><span class="header-section-number">4.3</span> Aggregate network</h2>
<p>We can get an aggregated communication network by counting the number of links between cell types or summarising the communication probabilities. These results are stored as two-dimensional matrices.</p>
<pre class="r"><code>cellchat &lt;- aggregateNet(cellchat)

pander(cellchat@net$count[1:5, 1:5])</code></pre>
<table style="width:72%;">
<colgroup>
<col width="16%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">FIB-A</th>
<th align="center">FIB-B</th>
<th align="center">FIB-C</th>
<th align="center">FIB-D</th>
<th align="center">FIB-E</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>FIB-A</strong></td>
<td align="center">5</td>
<td align="center">3</td>
<td align="center">6</td>
<td align="center">8</td>
<td align="center">8</td>
</tr>
<tr class="even">
<td align="center"><strong>FIB-B</strong></td>
<td align="center">4</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">2</td>
<td align="center">4</td>
</tr>
<tr class="odd">
<td align="center"><strong>FIB-C</strong></td>
<td align="center">5</td>
<td align="center">2</td>
<td align="center">4</td>
<td align="center">6</td>
<td align="center">8</td>
</tr>
<tr class="even">
<td align="center"><strong>FIB-D</strong></td>
<td align="center">12</td>
<td align="center">9</td>
<td align="center">15</td>
<td align="center">19</td>
<td align="center">21</td>
</tr>
<tr class="odd">
<td align="center"><strong>FIB-E</strong></td>
<td align="center">15</td>
<td align="center">13</td>
<td align="center">20</td>
<td align="center">25</td>
<td align="center">29</td>
</tr>
</tbody>
</table>
<pre class="r"><code>pander(cellchat@net$sum[1:5, 1:5])</code></pre>
<table style="width:100%;">
<colgroup>
<col width="16%" />
<col width="16%" />
<col width="16%" />
<col width="16%" />
<col width="16%" />
<col width="16%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">FIB-A</th>
<th align="center">FIB-B</th>
<th align="center">FIB-C</th>
<th align="center">FIB-D</th>
<th align="center">FIB-E</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>FIB-A</strong></td>
<td align="center">4.806e-06</td>
<td align="center">7.669e-07</td>
<td align="center">1.322e-06</td>
<td align="center">4.156e-06</td>
<td align="center">2.752e-06</td>
</tr>
<tr class="even">
<td align="center"><strong>FIB-B</strong></td>
<td align="center">1.435e-05</td>
<td align="center">3.489e-06</td>
<td align="center">1.862e-06</td>
<td align="center">2.748e-06</td>
<td align="center">2.677e-06</td>
</tr>
<tr class="odd">
<td align="center"><strong>FIB-C</strong></td>
<td align="center">4.64e-06</td>
<td align="center">2.509e-06</td>
<td align="center">1.434e-06</td>
<td align="center">3.339e-06</td>
<td align="center">2.769e-06</td>
</tr>
<tr class="even">
<td align="center"><strong>FIB-D</strong></td>
<td align="center">5.496e-05</td>
<td align="center">0.0001328</td>
<td align="center">7.748e-05</td>
<td align="center">0.0001942</td>
<td align="center">0.0001317</td>
</tr>
<tr class="odd">
<td align="center"><strong>FIB-E</strong></td>
<td align="center">2.67e-05</td>
<td align="center">7.208e-05</td>
<td align="center">4.105e-05</td>
<td align="center">0.0001038</td>
<td align="center">7.044e-05</td>
</tr>
</tbody>
</table>
<p class="timeit">
Chunk time: 0.05 secs
</p>
</div>
</div>
<div id="network-analysis" class="section level1" number="5">
<h1 number="5"><span class="header-section-number">5</span> Network analysis</h1>
<p><strong>CellChat</strong> can perform analysis on the communication network to better understand the roles of each cell type. Here we perform that analysis using the pathway scores. The output is a list with a set of metrics for cell type for each pathway.</p>
<pre class="r"><code>cellchat &lt;- netAnalysis_signalingRole(cellchat, slot.name = &quot;netP&quot;)</code></pre>
<p class="timeit">
Chunk time: 1.96 secs
</p>
<p>The network metrics are: <strong>outdeg</strong>, <strong>indeg</strong>, <strong>hub</strong>, <strong>authority</strong>, <strong>eigen</strong>, <strong>page_rank</strong>, <strong>betweenness</strong>, <strong>flowbet</strong> and <strong>info</strong></p>
</div>
<div id="global-patterns" class="section level1" number="6">
<h1 number="6"><span class="header-section-number">6</span> Global patterns</h1>
<p>It may also be interesting to identify common communication patterns. We can do this for both outgoing and incoming patterns.</p>
<pre class="r"><code>cellchat &lt;- identifyCommunicationPatterns(cellchat, pattern = &quot;outgoing&quot;, k = 5)
cellchat &lt;- identifyCommunicationPatterns(cellchat, pattern = &quot;incoming&quot;, k = 5)</code></pre>
<img src="figure/14-CellChat.Rmd/patterns-1.png" width="960" style="display: block; margin: auto;" />
<p class="timeit">
Chunk time: 7.9 secs
</p>
</div>
<div id="manifold-learning-and-classification" class="section level1" number="7">
<h1 number="7"><span class="header-section-number">7</span> Manifold learning and classification</h1>
<p><strong>CellChat</strong> can identify groups of signalling pathways by embedding and clustering them.</p>
<div id="functional-similarity" class="section level2" number="7.1">
<h2 number="7.1"><span class="header-section-number">7.1</span> Functional similarity</h2>
<p>Functional similarity indicates that major senders and receivers are similar and can be interpreted as the two pathways or interaction pairs exhibiting similar and/or redundant roles.</p>
<p>The output is a similarity matrix, UMAP embedding and cluster assignments from <span class="math inline">\(k\)</span>-means.</p>
<pre class="r"><code>cellchat &lt;- computeNetSimilarity(cellchat, type = &quot;functional&quot;, thresh = 0.25)
cellchat &lt;- netEmbedding(cellchat, type = &quot;functional&quot;)
cellchat &lt;- netClustering(cellchat, type = &quot;functional&quot;, k = 4)

pander(cellchat@netP$similarity$functional$matrix[1:5,1:5])</code></pre>
<table style="width:87%;">
<colgroup>
<col width="16%" />
<col width="15%" />
<col width="11%" />
<col width="15%" />
<col width="13%" />
<col width="15%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">TGFb</th>
<th align="center">ncWNT</th>
<th align="center">EGF</th>
<th align="center">PDGF</th>
<th align="center">IGF</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>TGFb</strong></td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">0.008282</td>
<td align="center">0.03183</td>
<td align="center">0.009311</td>
</tr>
<tr class="even">
<td align="center"><strong>ncWNT</strong></td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>EGF</strong></td>
<td align="center">0.008282</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">0.08182</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>PDGF</strong></td>
<td align="center">0.03183</td>
<td align="center">0</td>
<td align="center">0.08182</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>IGF</strong></td>
<td align="center">0.009311</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">1</td>
</tr>
</tbody>
</table>
<pre class="r"><code>pander(cellchat@netP$similarity$functional$dr[1:5, ])</code></pre>
<table style="width:44%;">
<colgroup>
<col width="16%" />
<col width="13%" />
<col width="13%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">UMAP1</th>
<th align="center">UMAP2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>TGFb</strong></td>
<td align="center">-2.057</td>
<td align="center">7.281</td>
</tr>
<tr class="even">
<td align="center"><strong>ncWNT</strong></td>
<td align="center">-0.6042</td>
<td align="center">10.15</td>
</tr>
<tr class="odd">
<td align="center"><strong>EGF</strong></td>
<td align="center">-0.9521</td>
<td align="center">10.94</td>
</tr>
<tr class="even">
<td align="center"><strong>PDGF</strong></td>
<td align="center">-1.836</td>
<td align="center">9.671</td>
</tr>
<tr class="odd">
<td align="center"><strong>IGF</strong></td>
<td align="center">-2.484</td>
<td align="center">4.831</td>
</tr>
</tbody>
</table>
<pre class="r"><code>cellchat@netP$similarity$functional$group</code></pre>
<pre><code>    TGFb    ncWNT      EGF     PDGF      IGF   APELIN      CCL     CXCL 
       3        4        4        4        2        1        1        1 
     MIF      IL1      CSF      TNF     SPP1   ANGPTL       MK      PTN 
       3        1        3        3        2        2        2        2 
     KIT    SEMA3      GAS GALECTIN     PROS 
       1        2        4        4        4 </code></pre>
<p class="timeit">
Chunk time: 13.36 secs
</p>
</div>
<div id="structural-similarity" class="section level2" number="7.2">
<h2 number="7.2"><span class="header-section-number">7.2</span> Structural similarity</h2>
<p>Structural similarity is used to compare the signalling network structure, without considering the similarity of senders and receivers.</p>
<pre class="r"><code>cellchat &lt;- computeNetSimilarity(cellchat, type = &quot;structural&quot;, thresh = 0.25)
cellchat &lt;- netEmbedding(cellchat, type = &quot;structural&quot;)
cellchat &lt;- netClustering(cellchat, type = &quot;structural&quot;, k = 4)

pander(cellchat@netP$similarity$structural$matrix[1:5,1:5])</code></pre>
<table style="width:76%;">
<colgroup>
<col width="16%" />
<col width="9%" />
<col width="12%" />
<col width="12%" />
<col width="12%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">TGFb</th>
<th align="center">ncWNT</th>
<th align="center">EGF</th>
<th align="center">PDGF</th>
<th align="center">IGF</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>TGFb</strong></td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>ncWNT</strong></td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">0.3642</td>
<td align="center">0.5772</td>
<td align="center">0.3458</td>
</tr>
<tr class="odd">
<td align="center"><strong>EGF</strong></td>
<td align="center">0</td>
<td align="center">0.3642</td>
<td align="center">1</td>
<td align="center">0.5978</td>
<td align="center">0.5932</td>
</tr>
<tr class="even">
<td align="center"><strong>PDGF</strong></td>
<td align="center">0</td>
<td align="center">0.5772</td>
<td align="center">0.5978</td>
<td align="center">1</td>
<td align="center">0.6122</td>
</tr>
<tr class="odd">
<td align="center"><strong>IGF</strong></td>
<td align="center">0</td>
<td align="center">0.3458</td>
<td align="center">0.5932</td>
<td align="center">0.6122</td>
<td align="center">1</td>
</tr>
</tbody>
</table>
<pre class="r"><code>pander(cellchat@netP$similarity$structural$dr[1:5, ])</code></pre>
<table style="width:39%;">
<colgroup>
<col width="16%" />
<col width="11%" />
<col width="11%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">UMAP1</th>
<th align="center">UMAP2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>TGFb</strong></td>
<td align="center">18.98</td>
<td align="center">4.604</td>
</tr>
<tr class="even">
<td align="center"><strong>ncWNT</strong></td>
<td align="center">29.88</td>
<td align="center">19.12</td>
</tr>
<tr class="odd">
<td align="center"><strong>EGF</strong></td>
<td align="center">30.36</td>
<td align="center">19.86</td>
</tr>
<tr class="even">
<td align="center"><strong>PDGF</strong></td>
<td align="center">30.5</td>
<td align="center">19.39</td>
</tr>
<tr class="odd">
<td align="center"><strong>IGF</strong></td>
<td align="center">29.47</td>
<td align="center">19.91</td>
</tr>
</tbody>
</table>
<pre class="r"><code>cellchat@netP$similarity$structural$group</code></pre>
<pre><code>    TGFb    ncWNT      EGF     PDGF      IGF   APELIN      CCL     CXCL 
       2        1        1        1        1        3        4        3 
     MIF      IL2      IL1      CSF      TNF    RANKL     SPP1   ANGPTL 
       2        3        4        1        1        3        4        2 
   ANGPT       MK      PTN      EDN      KIT    SEMA3      GAS GALECTIN 
       3        2        2        3        3        1        3        4 
    PROS 
       3 </code></pre>
<p class="timeit">
Chunk time: 11.7 secs
</p>
</div>
</div>
<div id="visualisation" class="section level1" number="8">
<h1 number="8"><span class="header-section-number">8</span> Visualisation</h1>
<div id="signalling-pathways" class="section level2" number="8.1">
<h2 number="8.1"><span class="header-section-number">8.1</span> Signalling pathways</h2>
<div id="hierarchy-plot" class="section level3" number="8.1.1">
<h3 number="8.1.1"><span class="header-section-number">8.1.1</span> Hierarchy plot</h3>
<p>Here solid circles are sources and open circles are targets. Circle sizes are the number of cells in each group and lines are coloured according to the source with thicker lines indicating stronger signals.</p>
<pre class="r"><code>netVisual_aggregate(cellchat, &quot;TGFb&quot;, vertex.receiver = 1:9, # groups to show
                    vertex.size = group_size)</code></pre>
<img src="figure/14-CellChat.Rmd/vis-pathway-hierarchy-1.png" width="960" style="display: block; margin: auto;" />
<p class="timeit">
Chunk time: 0.39 secs
</p>
</div>
<div id="circle-plot" class="section level3" number="8.1.2">
<h3 number="8.1.2"><span class="header-section-number">8.1.2</span> Circle plot</h3>
<p>We can show a similar thing as a circle plot.</p>
<pre class="r"><code>netVisual_aggregate(cellchat, &quot;ncWNT&quot;, layout = &quot;circle&quot;,
                    vertex.size = group_size)</code></pre>
<img src="figure/14-CellChat.Rmd/vis-pathway-circle-1.png" width="960" style="display: block; margin: auto;" />
<p class="timeit">
Chunk time: 0.2 secs
</p>
</div>
</div>
<div id="pathway-contribution" class="section level2" number="8.2">
<h2 number="8.2"><span class="header-section-number">8.2</span> Pathway contribution</h2>
<p>Here we show the contribution of each interaction pair to a signalling pathway.</p>
<pre class="r"><code>netAnalysis_contribution(cellchat, &quot;TGFb&quot;)</code></pre>
<img src="figure/14-CellChat.Rmd/vis-contribution-1.png" width="960" style="display: block; margin: auto;" />
<p class="timeit">
Chunk time: 0.35 secs
</p>
</div>
<div id="signalling-roles" class="section level2" number="8.3">
<h2 number="8.3"><span class="header-section-number">8.3</span> Signalling roles</h2>
<p>For can look at the expected role for each cell type for a pathway based on the network analysis scores.</p>
<pre class="r"><code>netVisual_signalingRole(cellchat, &quot;TGFb&quot;, width = 16, height = 4,
                        font.size = 10)</code></pre>
<img src="figure/14-CellChat.Rmd/vis-roles-1.png" width="960" style="display: block; margin: auto;" />
<p class="timeit">
Chunk time: 0.41 secs
</p>
</div>
<div id="communication-patterns" class="section level2" number="8.4">
<h2 number="8.4"><span class="header-section-number">8.4</span> Communication patterns</h2>
<div id="river-plot" class="section level3" number="8.4.1">
<h3 number="8.4.1"><span class="header-section-number">8.4.1</span> River plot</h3>
<p>An alternative way to visualise the communication patterns is using a river plot.</p>
<pre class="r"><code>netAnalysis_river(cellchat, pattern = &quot;outgoing&quot;)</code></pre>
<img src="figure/14-CellChat.Rmd/vis-patterns-river-1.png" width="960" style="display: block; margin: auto;" />
<p class="timeit">
Chunk time: 1.83 secs
</p>
</div>
<div id="dot-plot" class="section level3" number="8.4.2">
<h3 number="8.4.2"><span class="header-section-number">8.4.2</span> Dot plot</h3>
<p>Another alternative is a dot plot.</p>
<pre class="r"><code>netAnalysis_dot(cellchat, pattern = &quot;outgoing&quot;)</code></pre>
<img src="figure/14-CellChat.Rmd/vis-patterns-dot-1.png" width="960" style="display: block; margin: auto;" />
<p class="timeit">
Chunk time: 0.64 secs
</p>
</div>
</div>
<div id="embedding" class="section level2" number="8.5">
<h2 number="8.5"><span class="header-section-number">8.5</span> Embedding</h2>
<p>UMAP embedding of pathway similarity.</p>
<pre class="r"><code>netVisual_embedding(cellchat, type = &quot;functional&quot;, pathway.remove.show = FALSE,
                    label.size = 3.5)</code></pre>
<img src="figure/14-CellChat.Rmd/umap-1.png" width="960" style="display: block; margin: auto;" />
<p class="timeit">
Chunk time: 0.67 secs
</p>
<div id="zoom-in" class="section level3" number="8.5.1">
<h3 number="8.5.1"><span class="header-section-number">8.5.1</span> Zoom in</h3>
<pre class="r"><code>netVisual_embeddingZoomIn(cellchat, type = &quot;functional&quot;, nCol = 2)</code></pre>
<img src="figure/14-CellChat.Rmd/umap-zoom-1.png" width="960" style="display: block; margin: auto;" />
<p class="timeit">
Chunk time: 0.71 secs
</p>
</div>
</div>
</div>
<div id="summary" class="section level1 unnumbered" number="">
<h1 class="unnumbered" number="">Summary</h1>
<div id="parameters" class="section level2 unnumbered" number="">
<h2 class="unnumbered" number="">Parameters</h2>
<p>This table describes parameters used and set in this document.</p>
<pre class="r"><code>params &lt;- list(
    
)
params &lt;- toJSON(params, pretty = TRUE)
kable(fromJSON(params))</code></pre>
<table class="kable_wrapper">
<tbody>
<tr>
<td>
</td>
</tr>
</tbody>
</table>
<p class="timeit">
Chunk time: 0.01 secs
</p>
</div>
<div id="output-files" class="section level2 unnumbered" number="">
<h2 class="unnumbered" number="">Output files</h2>
<p>This table describes the output files produced by this document. Right click and <em>Save Link As…</em> to download the results.</p>
<pre class="r"><code>kable(data.frame(
    File = c(
        download_link(&quot;parameters.json&quot;, OUT_DIR)
    ),
    Description = c(
        &quot;Parameters set and used in this analysis&quot;
    )
))</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">File</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><a href="NA/raw/master/output/14-CellChat.Rmd/parameters.json">parameters.json</a></td>
<td align="left">Parameters set and used in this analysis</td>
</tr>
</tbody>
</table>
<p class="timeit">
Chunk time: 0.02 secs
</p>
</div>
<div id="session-information" class="section level2 unnumbered" number="">
<h2 class="unnumbered" number="">Session information</h2>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessioninfo::session_info()</code></pre>
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 4.0.0 (2020-04-24)
 os       macOS Catalina 10.15.6      
 system   x86_64, darwin17.0          
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       Europe/Berlin               
 date     2020-07-30                  

─ Packages ───────────────────────────────────────────────────────────────────
 ! package        * version date       lib
 P assertthat       0.2.1   2019-03-21 [?]
 P backports        1.1.6   2020-04-05 [?]
 P base64enc        0.1-3   2015-07-28 [?]
 P base64url        1.4     2018-05-14 [?]
   bibtex           0.4.2.2 2020-01-02 [1]
   Biobase        * 2.48.0  2020-04-27 [1]
   BiocGenerics   * 0.34.0  2020-04-27 [1]
 P broom            0.5.6   2020-04-20 [?]
 P CellChat       * 0.0.1   2020-07-29 [?]
 P cellranger       1.1.0   2016-07-27 [?]
 P circlize         0.4.9   2020-04-30 [?]
 P cli              2.0.2   2020-02-28 [?]
   clue             0.3-57  2019-02-25 [1]
 P cluster          2.1.0   2019-06-19 [3]
   coda             0.19-3  2019-07-05 [1]
 P codetools        0.2-16  2018-12-24 [3]
 P colorspace       1.4-1   2019-03-18 [?]
 P ComplexHeatmap   2.5.4   2020-07-29 [?]
 P conflicted     * 1.0.4   2019-06-21 [?]
 P cowplot          1.0.0   2019-07-11 [?]
 P crayon           1.3.4   2017-09-16 [?]
 P DBI              1.1.0   2019-12-15 [?]
 P dbplyr           1.4.3   2020-04-19 [?]
 P digest           0.6.25  2020-02-23 [?]
   doParallel       1.0.15  2019-08-02 [1]
 P dplyr          * 0.8.5   2020-03-07 [?]
 P drake            7.12.0  2020-03-25 [?]
 P ellipsis         0.3.0   2019-09-20 [?]
 P evaluate         0.14    2019-05-28 [?]
 P fansi            0.4.1   2020-01-08 [?]
 P farver           2.0.3   2020-01-16 [?]
 P filelock         1.0.2   2018-10-05 [?]
   FNN              1.1.3   2019-02-15 [1]
 P forcats        * 0.5.0   2020-03-01 [?]
   foreach          1.5.0   2020-03-30 [1]
 P fs             * 1.4.1   2020-04-04 [?]
 P future           1.17.0  2020-04-18 [?]
 P future.apply     1.5.0   2020-04-17 [?]
   gdtools          0.2.2   2020-04-03 [1]
 P generics         0.0.2   2018-11-29 [?]
 P GetoptLong       0.1.8   2020-01-08 [?]
   ggalluvial     * 0.12.0  2020-07-14 [1]
 P ggplot2        * 3.3.0   2020-03-05 [?]
   ggrepel          0.8.2   2020-03-08 [1]
 P git2r            0.27.1  2020-05-03 [?]
 P GlobalOptions    0.1.1   2019-09-30 [?]
   globals          0.12.5  2019-12-07 [1]
 P glue           * 1.4.0   2020-04-03 [?]
   gridBase         0.4-7   2014-02-24 [1]
 P gtable           0.3.0   2019-03-25 [?]
 P haven            2.2.0   2019-11-08 [?]
 P here           * 0.1     2017-05-28 [?]
 P highr            0.8     2019-03-20 [?]
 P hms              0.5.3   2020-01-08 [?]
 P htmltools        0.4.0   2019-10-04 [?]
 P httpuv           1.5.2   2019-09-11 [?]
 P httr             1.4.1   2019-08-05 [?]
 P igraph           1.2.5   2020-03-19 [?]
   IRanges          2.22.2  2020-05-21 [1]
   irlba            2.3.3   2019-02-05 [1]
   iterators        1.0.12  2019-07-26 [1]
 P jsonlite       * 1.6.1   2020-02-02 [?]
 P knitr          * 1.28    2020-02-06 [?]
 P labeling         0.3     2014-08-23 [?]
 P later            1.0.0   2019-10-04 [?]
 P lattice          0.20-41 2020-04-02 [?]
 P lifecycle        0.2.0   2020-03-06 [?]
   listenv          0.8.0   2019-12-05 [1]
 P lubridate        1.7.8   2020-04-06 [?]
   magick           2.4.0   2020-06-23 [3]
 P magrittr         1.5     2014-11-22 [?]
 P Matrix           1.2-18  2019-11-27 [?]
 P memoise          1.1.0   2017-04-21 [?]
 P modelr           0.1.7   2020-04-30 [?]
 P munsell          0.5.0   2018-06-12 [?]
   network          1.16.0  2019-12-01 [1]
 P nlme             3.1-147 2020-04-13 [?]
   NMF              0.22.0  2020-02-12 [1]
 P pander         * 0.6.3   2018-11-06 [?]
   pbapply          1.4-2   2019-08-31 [1]
 P pillar           1.4.4   2020-05-05 [?]
 P pkgconfig        2.0.3   2019-09-22 [?]
   pkgmaker         0.31.1  2020-03-19 [1]
 P plyr             1.8.6   2020-03-03 [?]
   png              0.1-7   2013-12-03 [1]
 P prettyunits      1.1.1   2020-01-24 [?]
 P progress         1.2.2   2019-05-16 [?]
 P promises         1.1.0   2019-10-04 [?]
 P purrr          * 0.3.4   2020-04-17 [?]
 P R6               2.4.1   2019-11-12 [?]
 P RColorBrewer     1.1-2   2014-12-07 [?]
 P Rcpp             1.0.4.6 2020-04-09 [?]
 P readr          * 1.3.1   2018-12-21 [?]
 P readxl           1.3.1   2019-03-13 [?]
   registry         0.5-1   2019-03-05 [1]
   renv             0.11.0  2020-06-26 [1]
 P repr             1.1.0   2020-01-28 [?]
 P reprex           0.3.0   2019-05-16 [?]
 P reshape2         1.4.4   2020-04-09 [?]
 P reticulate       1.16    2020-05-27 [?]
   rjson            0.2.20  2018-06-08 [1]
 P rlang            0.4.6   2020-05-02 [?]
 P rmarkdown        2.1     2020-01-20 [?]
   rngtools         1.5     2020-01-23 [1]
 P rprojroot        1.3-2   2018-01-03 [?]
   RSpectra         0.16-0  2019-12-01 [1]
 P rstudioapi       0.11    2020-02-07 [?]
 P rvest            0.3.5   2019-11-08 [?]
   S4Vectors        0.26.1  2020-05-16 [1]
 P scales           1.1.0   2019-11-18 [?]
 P sessioninfo      1.1.1   2018-11-05 [?]
 P shape            1.4.4   2018-02-07 [?]
 P skimr          * 2.1.1   2020-04-16 [?]
   sna              2.5     2019-12-10 [1]
   statnet.common   4.3.0   2019-06-02 [1]
 P storr            1.2.1   2018-10-18 [?]
 P stringi          1.4.6   2020-02-17 [?]
 P stringr        * 1.4.0   2019-02-10 [?]
   svglite          1.2.3.2 2020-07-07 [1]
   systemfonts      0.2.3   2020-06-09 [1]
 P tibble         * 3.0.1   2020-04-20 [?]
 P tidyr          * 1.0.3   2020-05-07 [?]
 P tidyselect       1.0.0   2020-01-27 [?]
 P tidyverse      * 1.3.0   2019-11-21 [?]
 P txtq             0.2.0   2019-10-15 [?]
 P vctrs            0.2.4   2020-03-10 [?]
 P whisker          0.4     2019-08-28 [?]
 P withr            2.2.0   2020-04-20 [?]
 P workflowr        1.6.2   2020-04-30 [?]
 P xfun             0.13    2020-04-13 [?]
 P xml2             1.3.2   2020-04-23 [?]
   xtable           1.8-4   2019-04-21 [1]
 P yaml             2.2.1   2020-02-01 [?]
 source                                  
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 standard (@1.4)                         
 CRAN (R 4.0.2)                          
 Bioconductor                            
 Bioconductor                            
 CRAN (R 4.0.0)                          
 Github (sqjin/CellChat@1df06c1)         
 standard (@1.1.0)                       
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.0)                          
 standard (@1.4-1)                       
 Github (jokergoo/ComplexHeatmap@d38254f)
 standard (@1.0.4)                       
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 standard (@0.14)                        
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.1)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.2)                          
 standard (@0.0.2)                       
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.2)                          
 standard (@0.3.0)                       
 standard (@2.2.0)                       
 standard (@0.1)                         
 standard (@0.8)                         
 CRAN (R 4.0.0)                          
 standard (@0.4.0)                       
 standard (@1.5.2)                       
 standard (@1.4.1)                       
 CRAN (R 4.0.0)                          
 Bioconductor                            
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 standard (@0.3)                         
 standard (@1.0.0)                       
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 standard (@1.2-18)                      
 standard (@1.1.0)                       
 CRAN (R 4.0.0)                          
 standard (@0.5.0)                       
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 standard (@1.1.0)                       
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 standard (@1.1-2)                       
 CRAN (R 4.0.0)                          
 standard (@1.3.1)                       
 standard (@1.3.1)                       
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.0)                          
 standard (@0.3.0)                       
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 standard (@0.3.5)                       
 Bioconductor                            
 standard (@1.1.0)                       
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.2)                          
 standard (@1.2.1)                       
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 standard (@1.3.0)                       
 standard (@0.2.0)                       
 CRAN (R 4.0.0)                          
 standard (@0.4)                         
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.0)                          
 CRAN (R 4.0.2)                          
 CRAN (R 4.0.0)                          

[1] /Users/luke.zappia/Documents/Projects/interaction-tools/renv/library/R-4.0/x86_64-apple-darwin17.0
[2] /private/var/folders/rj/60lhr791617422kqvh0r4vy40000gn/T/Rtmphp5zWa/renv-system-library
[3] /Library/Frameworks/R.framework/Versions/4.0/Resources/library

 P ── Loaded and on-disk path mismatch.</code></pre>
<p class="timeit">
Chunk time: 0.38 secs
</p>
</div>
</div>
</div>

<div id="rmd-source-code">---
title: "CellChat example"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

```{r setup, cache = FALSE}
# Setup document
source(here::here("code", "setup.R"))

# Function dependencies
invisible(drake::readd(download_link))
```

Introduction {.unnumbered}
============

In this document we are going to run through the example analysis for the
**CellChat** package and have a look at the output it produces. More information
about **CellChat** can be found at https://github.com/sqjin/CellChat.

```{r libraries}
library("CellChat")
library("ggalluvial")
```

Input
=====

**CellChat** requires two user inputs: an expression matrix and either cell-type
labels or a low-dimensional embedding. Here we use the label-based mode. The
other inputs to **CellChat** are the ligand-receptor database and
protein-protein interaction network included in the package.

```{r load}
load(fs::path(PATHS$CellChat_in, "data_wound_CellChat.rda"))
```

Expression matrix
-----------------

The expression matrix should be normalised with genes as rows and cells as
columns. Here is a snippet of the example dataset:

```{r input-expr}
expr <- data_wound$data

pander(as.matrix(expr[1:5, 1:5]))
```

The full matrix has **`r nrow(expr)`** rows (genes) and **`r ncol(expr)`**
columns (cells).

Labels
------

The second input is a `data.frame` with cell type labels for each cell.

```{r input-labels}
labels <- data.frame(
    group     = data_wound$labels,
    row.names = names(data_wound$labels)
)
group_size <- as.numeric(table(labels$group))

skim(labels)
```

Database
--------

The database is available as a list. The tutorial suggests using only the
interactions annotated with "Secreted Signalling".

```{r input-database}
CellChatDB <- CellChatDB.mouse
CellChatDB <- subsetDB(CellChatDB, search = "Secreted Signaling")
```

The database list contains **`r length(CellChatDB)`** items: 
`r glue_collapse(glue("**{names(CellChatDB)}**"), sep = ", ", last = " and ")`

### Interaction

The first item in the list is a `data.frame` with information about
interactions.

```{r input-database-interaction}
skim(CellChatDB$interaction)
```

### Complex

The second item contain information about complexes.

```{r input-database-complex}
skim(CellChatDB$complex)
```

### Cofactor

The next item contains information about cofactors.

```{r input-database-cofactor}
skim(CellChatDB$cofactor)
```

### Gene info

The final item in the database is gene annotation information.

```{r input-database-geneinfo}
skim(CellChatDB$geneInfo)
```

Network
-------

The network is a binary matrix showing interactions between proteins.

```{r input-network}
pander(as.matrix(PPI.mouse[1:5, 1:5]))
```

The full matrix has **`r nrow(PPI.mouse)`** rows and **`r ncol(PPI.mouse)`**
columns.

Create `CellChat` object
========================

**CellChat** uses a custom object which we can create from the expression matrix
and labels. We also store the database here.

```{r object}
cellchat <- createCellChat(expr)
cellchat <- addMeta(cellchat, meta = labels, meta.name = "labels")
cellchat <- setIdent(cellchat, ident.use = "labels")
cellchat@DB <- CellChatDB

cellchat
```

Pre-processing
==============

Subset signalling genes
-----------------------

First we subset the object to select only the signalling genes that are listed
in the database.

```{r pre-subset}
cellchat <- subsetData(cellchat)

cellchat
```

Identify over-expression
------------------------

The next step is to identify over-expressed genes and interactions.
Over-expressed genes are detected using a one-sided Wilcoxon test. Interactions
are selected if either a ligand or receptor is over-expressed.

```{r pre-over}
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)
```

This step has identified **`r length(cellchat@var.features)`** genes and
**`r nrow(cellchat@LR$LRsig)`** interactions.

Project data
------------

The gene expression data is then projected onto a protein-protein interaction
network.

```{r pre-project}
cellchat <- projectData(cellchat, PPI.mouse)
```

Communication network inference
===============================

Compute communication probability
---------------------------------

At this stage we can now compute the communication probability and infer the
communication network. The results are stored in the `net` slot of the
`CellChat` object which contains three-dimensional arrays of probabilities and
p-values for interactions between each pair of cell types for each interaction
pair. Here are snippets of these arrays for the first interaction pair.

```{r infer-probability}
cellchat <- computeCommunProb(cellchat)

pander(cellchat@net$prob[1:5, 1:5, 1])
pander(cellchat@net$pval[1:5, 1:5, 1])
```

Infer pathway signalling
------------------------

We also compute the communication probabilities at the pathway level by
summarising all ligand-receptor interactions associated with each pathway. The
output is similar but now instead of interaction pairs we have pathways.

```{r infer-pathway}
cellchat <- computeCommunProbPathway(cellchat)

pander(cellchat@netP$prob[1:5, 1:5, 1])
```

Aggregate network
-----------------

We can get an aggregated communication network by counting the number of links
between cell types or summarising the communication probabilities. These results
are stored as two-dimensional matrices.

```{r infer-aggregate}
cellchat <- aggregateNet(cellchat)

pander(cellchat@net$count[1:5, 1:5])
pander(cellchat@net$sum[1:5, 1:5])
```

Network analysis
================

**CellChat** can perform analysis on the communication network to better
understand the roles of each cell type. Here we perform that analysis using the
pathway scores. The output is a list with a set of metrics for cell type for
each pathway.

```{r network-analysis}
cellchat <- netAnalysis_signalingRole(cellchat, slot.name = "netP")
```

The network metrics are:
`r glue_collapse(glue("**{names(cellchat@netP$centr[[1]])}**"), sep = ", ", last = " and ")`

Global patterns
===============

It may also be interesting to identify common communication patterns. We can do
this for both outgoing and incoming patterns.

```{r patterns}
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "outgoing", k = 5)
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "incoming", k = 5)
```

Manifold learning and classification
====================================

**CellChat** can identify groups of signalling pathways by embedding and
clustering them.

Functional similarity
----------------------

Functional similarity indicates that major senders and receivers are similar
and can be interpreted as the two pathways or interaction pairs exhibiting
similar and/or redundant roles. 

The output is a similarity matrix, UMAP embedding and cluster assignments from
$k$-means.

```{r manifold-functional}
cellchat <- computeNetSimilarity(cellchat, type = "functional", thresh = 0.25)
cellchat <- netEmbedding(cellchat, type = "functional")
cellchat <- netClustering(cellchat, type = "functional", k = 4)

pander(cellchat@netP$similarity$functional$matrix[1:5,1:5])
pander(cellchat@netP$similarity$functional$dr[1:5, ])
cellchat@netP$similarity$functional$group
```

Structural similarity
---------------------

Structural similarity is used to compare the signalling network structure,
without considering the similarity of senders and receivers.

```{r manifold-structural}
cellchat <- computeNetSimilarity(cellchat, type = "structural", thresh = 0.25)
cellchat <- netEmbedding(cellchat, type = "structural")
cellchat <- netClustering(cellchat, type = "structural", k = 4)

pander(cellchat@netP$similarity$structural$matrix[1:5,1:5])
pander(cellchat@netP$similarity$structural$dr[1:5, ])
cellchat@netP$similarity$structural$group
```

Visualisation
=============

Signalling pathways
-------------------

### Hierarchy plot

Here solid circles are sources and open circles are targets. Circle sizes are
the number of cells in each group and lines are coloured according to the source
with thicker lines indicating stronger signals.

```{r vis-pathway-hierarchy}
netVisual_aggregate(cellchat, "TGFb", vertex.receiver = 1:9, # groups to show
                    vertex.size = group_size)
```

### Circle plot

We can show a similar thing as a circle plot.

```{r vis-pathway-circle}
netVisual_aggregate(cellchat, "ncWNT", layout = "circle",
                    vertex.size = group_size)
```

Pathway contribution
--------------------

Here we show the contribution of each interaction pair to a signalling pathway.

```{r vis-contribution}
netAnalysis_contribution(cellchat, "TGFb")
```

Signalling roles
----------------

For can look at the expected role for each cell type for a pathway based on the
network analysis scores.

```{r vis-roles}
netVisual_signalingRole(cellchat, "TGFb", width = 16, height = 4,
                        font.size = 10)
```

Communication patterns
----------------------

### River plot

An alternative way to visualise the communication patterns is using a river
plot.

```{r vis-patterns-river}
netAnalysis_river(cellchat, pattern = "outgoing")
```

### Dot plot

Another alternative is a dot plot.

```{r vis-patterns-dot}
netAnalysis_dot(cellchat, pattern = "outgoing")
```

Embedding
---------

UMAP embedding of pathway similarity.

```{r umap}
netVisual_embedding(cellchat, type = "functional", pathway.remove.show = FALSE,
                    label.size = 3.5)
```

### Zoom in

```{r umap-zoom}
netVisual_embeddingZoomIn(cellchat, type = "functional", nCol = 2)
```


Summary {.unnumbered}
=======

Parameters {.unnumbered}
----------

This table describes parameters used and set in this document.

```{r parameters}
params <- list(
    
)
params <- toJSON(params, pretty = TRUE)
kable(fromJSON(params))
```

Output files {.unnumbered}
------------

This table describes the output files produced by this document. Right click
and _Save Link As..._ to download the results.

```{r output}
kable(data.frame(
    File = c(
        download_link("parameters.json", OUT_DIR)
    ),
    Description = c(
        "Parameters set and used in this analysis"
    )
))
```

Session information {.unnumbered}
-------------------
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("14-CellChat.Rmd");
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
